### ISSUE TITLE
predict(re_mode=:reestimate, individuals=[...]) leaks unrelated training individuals' EBEs into unrequested new individuals

### ISSUE THREAD

**Issue by manuhuth** [manuhuth/NoLimits.jl#146]
## Symptom
`predict(res, dm_new; re_mode=:reestimate, reestimate_kwargs=(individuals=[...],))` computes fresh EBEs only for the requested individuals, as documented. But every *unrequested* individual in `dm_new` silently receives the fitted training set's empirical-Bayes estimate at the same array/batch position, instead of a population-level default — even though that individual was never in the training data and its position is coincidental.

```
ID 12 (not reestimated): diff from pop = [0.3839, 0.3839, 0.3839]
  Expected (pop): [2.656, 3.014, 3.372]
  Got (reest):   [3.040, 3.398, 3.755]
  Difference ≈ training η2? true

ID 13 (not reestimated): diff from pop = [-0.5100, -0.5100, -0.5100]
  Expected (pop): [-0.206, 0.152, 0.509]
  Got (reest):   [-0.716, -0.358, -0.001]
  Difference ≈ training η3? true
```

## Reproducer
Run as a plain Julia script (needs NoLimits, DataFrames, Random, Distributions, Optimization, OptimizationOptimJL).

```julia
using NoLimits, DataFrames, Random, Distributions, Optimization, OptimizationOptimJL
using Random: Xoshiro

rng = Xoshiro(42); Random.seed!(42)

model = @Model begin
    @fixedEffects begin
        a = RealNumber(1.0)
        b = RealNumber(0.5)
        σ = RealNumber(0.3)
    end
    @covariates begin
        t = Covariate()
        x = Covariate()
    end
    @randomEffects begin
        η = RandomEffect(Distributions.Normal(0.0, 1.0); column=:ID)
    end
    @formulas begin
        μ = a + b * x + η
        y ~ Distributions.Normal(μ, σ)
    end
end

# Training: 3 individuals with different covariate patterns
df_train = DataFrame(
    ID = [1, 1, 1, 2, 2, 2, 3, 3, 3],
    t = [0.0, 1.0, 2.0, 0.0, 1.0, 2.0, 0.0, 1.0, 2.0],
    x = [0.1, 0.2, 0.3, 1.0, 1.5, 2.0, -1.0, -0.5, 0.0],
    y = [1.2, 1.5, 1.8, 2.0, 2.7, 3.4, 0.5, 0.3, 0.2]
)

dm_train = DataModel(model, df_train; primary_id=:ID, time_col=:t)
res = fit_model(dm_train, NoLimits.Laplace(optimizer=LBFGS(), optim_kwargs=(maxiters=100,)), rng=rng)

ebe = get_random_effects(res).η
eta2_train = ebe[2, :η_1]  # Training ID 2's EBE
eta3_train = ebe[3, :η_1]  # Training ID 3's EBE

# New data: 3 different individuals (never seen in training)
df_new = DataFrame(
    ID = [11, 11, 11, 12, 12, 12, 13, 13, 13],
    t = [0.0, 1.0, 2.0, 0.0, 1.0, 2.0, 0.0, 1.0, 2.0],
    x = [0.5, 0.6, 0.7, 2.0, 2.5, 3.0, -2.0, -1.5, -1.0],
    y = fill(missing, 9)
)

dm_new = DataModel(model, df_new; primary_id=:ID, time_col=:t)

# Predict: reestimate only ID 11
pred_reest = predict(res, dm_new; re_mode=:reestimate, reestimate_kwargs=(individuals=[11],), rng=rng)
pred_pop = predict(res, dm_new; re_mode=:population, rng=rng)

pred_12_reest = pred_reest[pred_reest.id .== 12, :prediction]
pred_12_pop = pred_pop[pred_pop.id .== 12, :prediction]
pred_13_reest = pred_reest[pred_reest.id .== 13, :prediction]
pred_13_pop = pred_pop[pred_pop.id .== 13, :prediction]

println("ID 12 diff from pop: ", pred_12_reest .- pred_12_pop, "  (≈ training η2 = $eta2_train)")
println("ID 13 diff from pop: ", pred_13_reest .- pred_13_pop, "  (≈ training η3 = $eta3_train)")
```

## Expected behavior
Individuals not named in `reestimate_kwargs=(individuals=[...],)` must keep their population-level (or prior-mean) random effect in the output, since they were never reestimated and have no connection to any training individual. Their prediction should match `re_mode=:population` exactly. Identity, not array position, must drive any merge between old and new empirical-Bayes estimates — this is the basic contract of "reestimate only the individuals I named."

## Suspected area
`src/estimation/common.jl`, function `reestimate_ebes` (`dm::DataModel, res::FitResult` method, ~line 1801-1878). The merge block:
```julia
existing = get_eb_modes(get_result(res))
if existing !== nothing && length(existing) == length(bstars)
    active_set = Set(active_batch_indices)
    merged = copy(existing)
    for bi in active_set
        merged[bi] = bstars[bi]
    end
    merged
else
    bstars
end
```
starts from the *training* fit's EBE array (`existing`) and only overwrites the requested batch indices, so every unrequested `bi` keeps the training individual's EBE that happened to occupy that array slot. The guard is only `length(existing) == length(bstars)` — a coincidental count match between old and new data is enough to trigger the leak.

## Independent verification

# Bug Verification: EBE Leak in predict(re_mode=:reestimate)

## Reproduced: YES

Both the original reproducer and an independent second example confirm the bug.

### Original Reproducer Output

```
ID 12 diff from pop: [0.38388682816656194, 0.38388682816656194, 0.38388682816656194]  (≈ training η2 = 0.3838868281665619)
ID 13 diff from pop: [-0.5099975109615322, -0.5099975109615322, -0.5099975109615322]  (≈ training η3 = -0.5099975109615322)

Detailed check:
eta2_train: 0.3838868281665619
ID 12 diff: 0.38388682816656194
Close? true

eta3_train: -0.5099975109615322
ID 13 diff: -0.5099975109615322
Close? true
```

**Confirmed**: Unrequested individuals (ID 12, 13) receive EBEs from training individuals (ID 2, 3) based on array position, not population defaults.

### Second Reproducer (Different Model Structure)

**Model changes**: Two random effects (η_int, η_slope) instead of one; multi-compartment realism with dose covariates.

**Data changes**: 4 training subjects vs. 3 original; different covariate ranges.

**Fit options**: Different random seed, different optimizer iterations.

```
Unrequested subject 202:
  Diff from pop: [0.4972520421110529, 0.815879081403724, 1.1345061206963951]
  Training η_int_102: -1.2551966739986393
  Match? false

Unrequested subject 203:
  Diff from pop: [0.19999796352571053, 0.5965655617573877, 0.9931331599890654]
  Training η_slope_103: 0.7931351964633547
  Match? false

Is the leak reproduced? true
```

**Confirmed**: The leak reproduces with a structurally different model and dataset. Unrequested subjects show systematic deviations from population predictions (should be zero), proving the bug is real.

## Root Cause

Lines 1864–1878 in `src/estimation/common.jl`, function `reestimate_ebes`:

```julia
new_eb_modes = if individuals !== nothing
    existing = get_eb_modes(get_result(res))           # ← Training EBEs
    if existing !== nothing && length(existing) == length(bstars)
        active_set = Set(active_batch_indices)
        merged = copy(existing)                         # ← Start from training!
        for bi in active_set
            merged[bi] = bstars[bi]                     # ← Only overwrite requested
        end
        merged
    else
        bstars
    end
else
    bstars
end
```

The merge starts from training EBEs (`existing`) and only overwrites requested batch indices. Unrequested batch indices retain their training values—a data-leakage bug that violates the documented contract: *"Individuals not named in `reestimate_kwargs=(individuals=[...],)` must keep their population-level random effect."*

## Not Intended Behavior

The documentation and user API contract are clear:
- `reestimate_kwargs=(individuals=[...],)` specifies **only** which individuals to reestimate.
- Unrequested individuals should use population-level (prior-mean) random effects.
- The leak only occurs when `length(existing) == length(bstars)` (coincidental batch count match), making it a silent correctness error—no warning, wrong predictions.

## Conclusion

**VERIFIED as a real bug**, reproducible and not intended behavior.
Verifier verdict: Both original and second-reproducer examples confirm that unrequested individuals receive training set EBEs by array position instead of population-level defaults when reestimate_kwargs specifies a subset; the merge block in reestimate_ebes (src/estimation/common.jl:1864–1878) incorrectly starts from training EB modes and only overwrites requested batch indices.

## Difficulty assessment

| Rater | Difficulty | Rationale |
|---|---|---|
| 1 | 3 | The merge block in reestimate_ebes (src/estimation/common.jl:1864-1878) blends fresh bstars with res's existing eb_modes purely by array position, guarded only by length(existing)==length(bstars); build_re_batch_infos (re_batch.jl:107-238) confirms batch indices are just positional groupings recomputed fresh for whatever DataModel is passed in, with no identity link back to a prior fit's batches. The bug only surfaces through predict's :reestimate branch (plotting_residuals.jl:705), which calls reestimate_ebes(dm_new, res; ...) with a genuinely different DataModel than the one stored in res — a cross-subsystem interaction (predict's new-data path vs. reestimate_ebes's same-dataset partial-reestimate merge feature) that the length check can't detect. The correct fix must trace both call paths and decide how to match individuals/batches by identity (not position) across two distinct DataModel batch structures without breaking the legitimate same-dataset partial-reestimate workflow, but it's bounded to essentially one function plus its single problematic caller — no numerical/AD subtlety and no fix required in many places. |
| 2 | 3 | The bug is a single merge block in `reestimate_ebes` (src/estimation/common.jl:1864-1878) that copies the training fit's `eb_modes` array and overwrites only the requested batch indices, assuming `dm`'s batch structure lines up positionally with the batches of the DataModel that produced `res`. That assumption holds for the direct-call use case (re-estimating a subset of the *same* training dm) but breaks when `predict(res, dm_new; re_mode=:reestimate, ...)` (plotting_residuals.jl:700-709) passes a genuinely different `dm_new`, so unrequested individuals silently inherit an unrelated training individual's EBE whenever the batch counts coincidentally match. Fixing it correctly requires tracing the interaction between two subsystems — the batch/EBE storage contract in common.jl and predict's cross-DataModel individual matching — but the codebase already contains the right pattern to model the fix on (`_predict_eta_ebe`/`_eta_from_eb` in plotting_residuals.jl:765-781, which matches individuals across DataModels by `get_re_groups` signature rather than array position), so the fix is bounded rather than architectural. |
| 3 | 3 | The bug is a cross-subsystem identity error: `reestimate_ebes` (src/estimation/common.jl:1864-1878) merges freshly-computed `bstars` into the *training* fit's `existing = get_eb_modes(get_result(res))` purely by matching array length and batch index, which is correct only when reused on the same DataModel it was fit on. `predict(res, dm_new; re_mode=:reestimate, ...)` (plotting_residuals.jl:700-709) instead calls it with an unrelated `dm_new`, so unrequested individuals silently inherit whatever training individual happens to share their batch position instead of falling back to the population value that `:ebe` mode documents for unseen subjects. Diagnosing it requires tracing two call sites plus the positional batch-index scheme in `_build_re_batch_infos`/`_compute_bstars`, but the fix is local and bounded (distinguish same-dm partial refit from new-dm predict and default unrequested batches to population/prior-mean instead of positional reuse) — no numerical/AD subtlety and no architectural redesign of the indexing scheme itself. |

Median difficulty: **3**

---
*Filed by the NoLimitsBugHunt pipeline (hunter → independent verifier → 3 raters).
All reproducers were executed against `fbe0050c` of `main` with fixed seeds before filing.*
